batcher/aligned-batcher/src/lib.rs

```diff
@@ -84,6 +92,99 @@ impl BatchState {
             .or_insert(1);
     }
+
+    fn get_entry(&self, sender: Address, nonce: U256) -> Option<&BatchQueueEntry> {
+        self.batch_queue
+            .iter()
+            .map(|(entry, _)| entry)
+            .find(|entry| {
+                entry.sender == sender
+                    && U256::from_big_endian(entry.nonced_verification_data.nonce.as_slice())
+                        == nonce
+            })
+    }
+
+    /// Checks if the entry is valid
+    /// An entry is valid if there is no entry with the same sender,
+    /// lower nonce and a lower fee
+    /// If the entry is valid, it replaces the entry in the queue
+    /// to increment the max fee, then it updates the user min fee if necessary
+    /// If the entry is invalid, it returns a validity response message.
+    /// If the entry is valid, it returns None.
+    fn validate_and_increment_max_fee(
+        &mut self,
+        replacement_entry: BatchQueueEntry,
+    ) -> Option<ValidityResponseMessage> {
+        let replacement_max_fee = replacement_entry.nonced_verification_data.max_fee;
+        let nonce =
+            U256::from_big_endian(replacement_entry.nonced_verification_data.nonce.as_slice());
+        let sender = replacement_entry.sender;
+
+        debug!(
+            "Checking validity of entry with sender: {:?}, nonce: {:?}, max_fee: {:?}",
+            sender, nonce, replacement_max_fee
+        );
+
+        // it is a valid entry only if there is no entry with the same sender, lower nonce and a lower fee
+        let is_valid = !self.batch_queue.iter().any(|(entry, _)| {
+            entry.sender == sender
+                && U256::from_big_endian(entry.nonced_verification_data.nonce.as_slice()) < nonce
+                && entry.nonced_verification_data.max_fee < replacement_max_fee
+        });
+
+        if !is_valid {
+            return Some(ValidityResponseMessage::InvalidReplacementMessage);
+        }
+
+        info!(
+            "Entry is valid, incrementing fee for sender: {:?}, nonce: {:?}, max_fee: {:?}",
+            sender, nonce, replacement_max_fee
+        );
+
+        // remove the old entry and insert the new one
+        // note that the entries are considered equal for the priority queue
+        // if they have the same nonce and sender, so we can remove the old entry
+        // by calling remove with the new entry
+        self.batch_queue.remove(&replacement_entry);
+        self.batch_queue.push(
+            replacement_entry.clone(),
+            BatchQueueEntryPriority::new(replacement_max_fee, nonce),
+        );
+
+        let user_min_fee = self
+            .batch_queue
+            .iter()
+            .filter(|(e, _)| e.sender == sender)
+            .map(|(e, _)| e.nonced_verification_data.max_fee)
+            .min()
+            .unwrap_or(U256::max_value());
+
+        self.user_min_fee.insert(sender, user_min_fee);
+
+        None
+    }
+
+    fn update_user_proofs_in_batch_and_min_fee(&mut self) {
```

Comment:
since the whole batch may not be included in the finalized batch, we need to update the cache for:
- The user proof count in batch
- The user min fee pending in batch (which is the one with the highest nonce)

This caches are needed to validate user messages (balance and max fee)

Reply:
Could you add this comment to the function?

Reply:
added on 9d72137e66eebbe631e68dc0e82dee260d10bf18